// @metamask/json-rpc-engine/dist/v2/JsonRpcServer.mjs
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _a, _JsonRpcServer_engine, _JsonRpcServer_onError, _JsonRpcServer_coerceRequest;
import { rpcErrors, serializeError } from "@metamask/rpc-errors";
import { hasProperty, isObject } from "@metamask/utils";
import { JsonRpcEngineV2 } from "./JsonRpcEngineV2.mjs";
import { getUniqueId } from "../getUniqueId.mjs";
const jsonrpc = '2.0';
/**
 * A JSON-RPC server that handles requests and notifications.
 *
 * Essentially wraps a {@link JsonRpcEngineV2} in order to create a conformant
 * yet permissive JSON-RPC 2.0 server.
 *
 * Note that the server will accept both requests and notifications via {@link handle},
 * even if the underlying engine is only able to handle one or the other.
 *
 * @example
 * ```ts
 * const server = new JsonRpcServer({
 *   engine,
 *   onError,
 * });
 *
 * const response = await server.handle(request);
 * if ('result' in response) {
 *   // Handle result
 * } else {
 *   // Handle error
 * }
 * ```
 */
export class JsonRpcServer {
    /**
     * Construct a new JSON-RPC server.
     *
     * @param options - The options for the server.
     * @param options.onError - The callback to handle errors thrown by the
     * engine. Errors always result in a failed response object, containing a
     * JSON-RPC 2.0 serialized version of the original error. If you need to
     * access the original error, use the `onError` callback.
     * @param options.engine - The engine to use. Mutually exclusive with
     * `middleware`.
     * @param options.middleware - The middleware to use. Mutually exclusive with
     * `engine`.
     */
    constructor(options) {
        _JsonRpcServer_engine.set(this, void 0);
        _JsonRpcServer_onError.set(this, void 0);
        __classPrivateFieldSet(this, _JsonRpcServer_onError, options.onError, "f");
        if (hasProperty(options, 'engine')) {
            // @ts-expect-error - hasProperty fails to narrow the type.
            __classPrivateFieldSet(this, _JsonRpcServer_engine, options.engine, "f");
        }
        else {
            // @ts-expect-error - TypeScript complains that engine is of the wrong type, but clearly it's not.
            __classPrivateFieldSet(this, _JsonRpcServer_engine, JsonRpcEngineV2.create({ middleware: options.middleware }), "f");
        }
    }
    async handle(rawRequest, options) {
        // If rawRequest is not a notification, the originalId will be attached
        // to the response. We attach our own, trusted id in #coerceRequest()
        // while the request is being handled.
        const [originalId, isRequest] = getOriginalId(rawRequest);
        try {
            const request = __classPrivateFieldGet(_a, _a, "m", _JsonRpcServer_coerceRequest).call(_a, rawRequest, isRequest);
            // @ts-expect-error - The request may not be of the type expected by the engine,
            // and we intentionally allow this to happen.
            const result = await __classPrivateFieldGet(this, _JsonRpcServer_engine, "f").handle(request, options);
            if (result !== undefined) {
                return {
                    jsonrpc,
                    // @ts-expect-error - Reassign the original id, regardless of its type.
                    id: originalId,
                    result,
                };
            }
        }
        catch (error) {
            __classPrivateFieldGet(this, _JsonRpcServer_onError, "f")?.call(this, error);
            if (isRequest) {
                return {
                    jsonrpc,
                    // @ts-expect-error - Reassign the original id, regardless of its type.
                    id: originalId,
                    error: serializeError(error, {
                        shouldIncludeStack: false,
                        shouldPreserveMessage: true,
                    }),
                };
            }
        }
        return undefined;
    }
}
_a = JsonRpcServer, _JsonRpcServer_engine = new WeakMap(), _JsonRpcServer_onError = new WeakMap(), _JsonRpcServer_coerceRequest = function _JsonRpcServer_coerceRequest(rawRequest, isRequest) {
    if (!isMinimalRequest(rawRequest)) {
        throw rpcErrors.invalidRequest({
            data: {
                request: rawRequest,
            },
        });
    }
    const request = {
        jsonrpc,
        method: rawRequest.method,
    };
    if (hasProperty(rawRequest, 'params')) {
        request.params = rawRequest.params;
    }
    if (isRequest) {
        request.id = getUniqueId();
    }
    return request;
};
/**
 * Check if an unvalidated request is a minimal request.
 *
 * @param rawRequest - The raw request to check.
 * @returns `true` if the request is a {@link MinimalRequest}, `false` otherwise.
 */
function isMinimalRequest(rawRequest) {
    return (isObject(rawRequest) &&
        hasProperty(rawRequest, 'method') &&
        typeof rawRequest.method === 'string' &&
        hasValidParams(rawRequest));
}
/**
 * Check if a request has valid params, i.e. an array or object.
 * The contents of the params are not inspected.
 *
 * @param rawRequest - The request to check.
 * @returns `true` if the request has valid params, `false` otherwise.
 */
function hasValidParams(rawRequest) {
    if (hasProperty(rawRequest, 'params')) {
        return Array.isArray(rawRequest.params) || isObject(rawRequest.params);
    }
    return true;
}
/**
 * Get the original id from a request.
 *
 * @param rawRequest - The request to get the original id from.
 * @returns The original id and a boolean indicating if the request is a request
 * (as opposed to a notification).
 */
function getOriginalId(rawRequest) {
    if (isObject(rawRequest) && hasProperty(rawRequest, 'id')) {
        return [rawRequest.id, true];
    }
    return [undefined, false];
}
